providers/implementations/keymgmt/mlkem_kmgmt.c

```diff
@@ -0,0 +1,575 @@
+        gctx->provctx = provctx;
+        gctx->selection = selection;
+    }
+    if (!mlkem_gen_set_params(gctx, params)) {
+        OPENSSL_free(gctx);
+        gctx = NULL;
+    }
+    debug_print("MLKEMKM gen init returns %p\n", gctx);
+    return gctx;
+}
+
+static int mlkem_gen_set_params(void *genctx, const OSSL_PARAM params[])
+{
+    struct mlkem_gen_ctx *gctx = genctx;
+
+    if (gctx == NULL)
+        return 0;
+
+    debug_print("MLKEMKM empty gen_set params called for %p\n", gctx);
+    return 1;
+}
+
+static const OSSL_PARAM *mlkem_gen_settable_params(ossl_unused void *genctx,
+                                                   ossl_unused void *provctx)
+{
+    static OSSL_PARAM settable[] = {
+        OSSL_PARAM_END
+    };
+    return settable;
+}
+
+static void *mlkem_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
+{
+    struct mlkem_gen_ctx *gctx = (struct mlkem_gen_ctx *)vctx;
+    MLKEM768_KEY *mkey;
+
+    debug_print("MLKEMKM gen called for %p\n", gctx);
+    if (gctx == NULL)
+        return NULL;
+
+    if ((mkey = mlkem_new(NULL)) == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_INTERNAL_ERROR);
+        return NULL;
+    }
+
+    /* If we're doing parameter generation then we just return a blank key */
+    if ((gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0) {
+        debug_print("MLKEMKM gen returns blank %p\n", mkey);
+        return mkey;
+    }
+
+    mkey->keytype = MLKEM_KEY_TYPE_768;
+
+    if (mkey->encoded_pubkey == NULL) {
+        mkey->encoded_pubkey = OPENSSL_malloc(OSSL_MLKEM768_PUBLIC_KEY_BYTES);
+        if (mkey->encoded_pubkey == NULL) {
+            OPENSSL_free(mkey);
+            return NULL;
+        }
+    }
+
+    ossl_mlkem768_generate_key(mkey->encoded_pubkey, NULL, &mkey->seckey);
+    mkey->seckey_initialized = 1;
+    ossl_mlkem768_public_from_private(&mkey->pubkey, &mkey->seckey);
+    mkey->pubkey_initialized = 1;
+
+    debug_print("MLKEMKM gen returns set %p\n", mkey);
+    return mkey;
+}
+
+static void mlkem_gen_cleanup(void *genctx)
+{
+    struct mlkem_gen_ctx *gctx = genctx;
+
+    debug_print("MLKEMKM gen cleanup for %p\n", gctx);
+    OPENSSL_free(gctx);
+}
+
+static void *mlkem_dup(const void *vsrckey, int selection)
+{
+    const MLKEM768_KEY *srckey = (const MLKEM768_KEY *)vsrckey;
+    MLKEM768_KEY *dstkey;
+
+    debug_print("MLKEMKM dup called for %p\n", srckey);
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    dstkey = mlkem_new(NULL);
+    if (dstkey == NULL)
+        return NULL;
+
+    dstkey->keytype = srckey->keytype;
+    if (srckey->pubkey_initialized == 1
+            && (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
+        /* TODO(ML-KEM)/TBC: By commenting out these, the EVP level test still passes: WHY?? */
+        memcpy((void *)&dstkey->pubkey, (void *)&srckey->pubkey, sizeof(srckey->pubkey));
+        dstkey->encoded_pubkey = OPENSSL_malloc(OSSL_MLKEM768_PUBLIC_KEY_BYTES);
+        assert(srckey->encoded_pubkey != NULL);
```

Comment:
This should not be an assert. Just a simple `if (srckey->encoded_pubkey == NULL)`

Reply:
so changed in https://github.com/openssl/openssl/commit/467e70f9af6925357f8f418f4e47e354c0a9fa15